File: Android/BlueMessenger/app/src/main/java/com/omercankoc/bluemessenger/MainActivity.java
package com.omercankoc.bluemessenger;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private Context context;
    private BluetoothAdapter bluetoothAdapter;
    private ChatClass chatClass;

    private EditText editTextMessage;
    private Button buttonSend;
    private ListView listViewMessages;
    private ArrayAdapter<String> adapterMessages;

    private final int LOCATION_PERMISSION_REQUEST = 101;
    private final int SELECT_DEVICE = 102;

    public static final int MESSAGE_STATE_CHANGED = 0;
    public static final int MESSAGE_READ = 1;
    public static final int MESSAGE_WRITE = 2;
    public static final int MESSAGE_DEVICE_NAME = 3;
    public static final int MESSAGE_TOAST = 4;

    public static final String DEVICE_NAME = "device_name";
    public static final String TOAST = "toast";
    private String connectedDevice;

    private Handler handler = new Handler(new Handler.Callback() {
        @Override
        public boolean handleMessage(Message message) {
            switch (message.what){
                case MESSAGE_STATE_CHANGED:
                    switch (message.arg1){
                        case ChatClass.STATE_NONE:
                            setState("Not Connected!");
                            break;
                        case ChatClass.STATE_LISTEN:
                            setState("Not Connected...");
                            break;
                        case ChatClass.STATE_CONNECTING:
                            setState("Connecting...");
                            break;
                        case ChatClass.STATE_CONNECTED:
                            setState("Connected " + connectedDevice);
                            break;
                    }
                case MESSAGE_WRITE:
                    byte[] buffer1 = (byte[]) message.obj;
                    String outputBuffer = new String(buffer1);
                    adapterMessages.add("Me: " + outputBuffer);
                    break;
                case MESSAGE_READ:
                    byte[] buffer = (byte[]) message.obj;
                    String inputBuffer = new String(buffer, 0, message.arg1);
                    adapterMessages.add(connectedDevice + ": " + inputBuffer);
                    break;
                case MESSAGE_DEVICE_NAME:
                    connectedDevice = message.getData().getString(DEVICE_NAME);
                    Toast.makeText(context,connectedDevice,Toast.LENGTH_LONG).show();
                    break;
                case MESSAGE_TOAST:
                    Toast.makeText(context,message.getData().getString(TOAST),Toast.LENGTH_LONG).show();
                    break;
            }

            return false;
        }
    });

    private void setState(CharSequence subTitle){
        getSupportActionBar().setSubtitle(subTitle);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        context = this;

        initialize();
        initBluetooth();

        chatClass = new ChatClass(context,handler);
    }

    private void initialize(){
        editTextMessage = findViewById(R.id.editTextMessage);
        buttonSend = findViewById(R.id.buttonSend);
        listViewMessages = findViewById(R.id.listViewMessages);
        adapterMessages = new ArrayAdapter<String>(context,R.layout.message_layout);
        listViewMessages.setAdapter(adapterMessages);

        buttonSend.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String message = editTextMessage.getText().toString();
                if(message.isEmpty()){
                    editTextMessage.setText("");
                    chatClass.write(message.getBytes());
                }
            }
        });
    }

    /*

     */
    private void initBluetooth(){
        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        if(bluetoothAdapter == null){
            Toast.makeText(context,"No Bluetooth Found!",Toast.LENGTH_LONG).show();
        }
    }

    /*

     */
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main_activity,menu);
        return super.onCreateOptionsMenu(menu);
    }

    /*

     */
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.menu_search_devices:
                checkPermission();
                return true;
            case R.id.menu_bluetooth_on_off:
                enableBluetooth();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    /*

     */
    private void checkPermission(){
        if(ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED){
            ActivityCompat.requestPermissions(MainActivity.this,
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},LOCATION_PERMISSION_REQUEST);
        } else {
            Intent intent = new Intent(context,DeviceListActivity.class);
            startActivityForResult(intent,SELECT_DEVICE);
        }
    }


    /*
    DeviceListActivity'den Intent ile gönderilen Device Address bilgisini al.
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if(requestCode == SELECT_DEVICE && resultCode == RESULT_OK){
            String address = data.getStringExtra("deviceAddress");
            chatClass.connect(bluetoothAdapter.getRemoteDevice(address));
        }

        super.onActivityResult(requestCode, resultCode, data);
    }



    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if(requestCode == LOCATION_PERMISSION_REQUEST){
            if(grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
                Intent intent = new Intent(context,DeviceListActivity.class);
                startActivityForResult(intent,SELECT_DEVICE);
            } else {
                new AlertDialog.Builder(context)
                        .setCancelable(false)
                        .setMessage("Location Permission is Required! \n Please Grant!")
                        .setPositiveButton("Grant", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                checkPermission();
                            }
                        })
                        .setNegativeButton("Deny", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                MainActivity.this.finish();
                            }
                        })
                        .show();
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    /*

     */
    private void enableBluetooth(){
        if (bluetoothAdapter.isEnabled()) {
            bluetoothAdapter.enable();
        }

        if(bluetoothAdapter.getScanMode() != BluetoothAdapter.SCAN_MODE_CONNECTABLE_DISCOVERABLE){
            Intent discoveryIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_DISCOVERABLE);
            discoveryIntent.putExtra(BluetoothAdapter.EXTRA_DISCOVERABLE_DURATION,300);
            startActivity(discoveryIntent);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if(chatClass != null){
            chatClass.stop();
        }
    }
}

File: Android/BlueMessenger/app/src/main/java/com/omercankoc/bluemessenger/MainActivity.kt
package com.omercankoc.bluemessenger

import android.Manifest
import androidx.appcompat.app.AppCompatActivity
import android.bluetooth.BluetoothAdapter
import com.omercankoc.bluemessenger.ChatClass
import com.omercankoc.bluemessenger.MainActivity
import android.os.Bundle
import com.omercankoc.bluemessenger.R
import androidx.core.content.ContextCompat
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import android.content.Intent
import com.omercankoc.bluemessenger.DeviceListActivity
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.os.Handler
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog

class MainActivity : AppCompatActivity() {
    private var context: Context? = null
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var chatClass: ChatClass? = null
    private var editTextMessage: EditText? = null
    private var buttonSend: Button? = null
    private var listViewMessages: ListView? = null
    private var adapterMessages: ArrayAdapter<String>? = null
    private val LOCATION_PERMISSION_REQUEST = 101
    private val SELECT_DEVICE = 102
    private var connectedDevice: String? = null
    private val handler = Handler { message ->
        when (message.what) {
            MESSAGE_STATE_CHANGED -> {
                when (message.arg1) {
                    ChatClass.STATE_NONE -> setState("Not Connected!")
                    ChatClass.STATE_LISTEN -> setState("Not Connected...")
                    ChatClass.STATE_CONNECTING -> setState("Connecting...")
                    ChatClass.STATE_CONNECTED -> setState("Connected $connectedDevice")
                }
                val buffer1 = message.obj as ByteArray
                val outputBuffer = String(buffer1)
                adapterMessages!!.add("Me: $outputBuffer")
            }
            MESSAGE_WRITE -> {
                val buffer1 = message.obj as ByteArray
                val outputBuffer = String(buffer1)
                adapterMessages!!.add("Me: $outputBuffer")
            }
            MESSAGE_READ -> {
                val buffer = message.obj as ByteArray
                val inputBuffer = String(buffer, 0, message.arg1)
                adapterMessages!!.add("$connectedDevice: $inputBuffer")
            }
            MESSAGE_DEVICE_NAME -> {
                connectedDevice = message.data.getString(DEVICE_NAME)
                Toast.makeText(context, connectedDevice, Toast.LENGTH_LONG).show()
            }
            MESSAGE_TOAST -> Toast.makeText(context, message.data.getString(TOAST), Toast.LENGTH_LONG).show()
        }
        false
    }

    private fun setState(subTitle: CharSequence) {
        supportActionBar!!.subtitle = subTitle
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        context = this
        initialize()
        initBluetooth()
        chatClass = ChatClass(context, handler)
    }

    private fun initialize() {
        editTextMessage = findViewById(R.id.editTextMessage)
        buttonSend = findViewById(R.id.buttonSend)
        listViewMessages = findViewById(R.id.listViewMessages)
        adapterMessages = ArrayAdapter(context!!, R.layout.message_layout)
        listViewMessages.setAdapter(adapterMessages)
        buttonSend.setOnClickListener(View.OnClickListener {
            val message = editTextMessage.getText().toString()
            if (message.isEmpty()) {
                editTextMessage.setText("")
                chatClass!!.write(message.toByteArray())
            }
        })
    }

    /*

     */
    private fun initBluetooth() {
        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        if (bluetoothAdapter == null) {
            Toast.makeText(context, "No Bluetooth Found!", Toast.LENGTH_LONG).show()
        }
    }

    /*

     */
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main_activity, menu)
        return super.onCreateOptionsMenu(menu)
    }

    /*

     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_search_devices -> {
                checkPermission()
                true
            }
            R.id.menu_bluetooth_on_off -> {
                enableBluetooth()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    /*

     */
    private fun checkPermission() {
        if (ContextCompat.checkSelfPermission(context!!, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), LOCATION_PERMISSION_REQUEST)
        } else {
            val intent = Intent(context, DeviceListActivity::class.java)
            startActivityForResult(intent, SELECT_DEVICE)
        }
    }

    /*
    DeviceListActivity'den Intent ile gönderilen Device Address bilgisini al.
     */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == SELECT_DEVICE && resultCode == RESULT_OK) {
            val address = data!!.getStringExtra("deviceAddress")
            chatClass!!.connect(bluetoothAdapter!!.getRemoteDevice(address))
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == LOCATION_PERMISSION_REQUEST) {
            if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                val intent = Intent(context, DeviceListActivity::class.java)
                startActivityForResult(intent, SELECT_DEVICE)
            } else {
                AlertDialog.Builder(context!!)
                        .setCancelable(false)
                        .setMessage("Location Permission is Required! \n Please Grant!")
                        .setPositiveButton("Grant") { dialogInterface, i -> checkPermission() }
                        .setNegativeButton("Deny") { dialogInterface, i -> finish() }
                        .show()
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    /*

     */
    private fun enableBluetooth() {
        if (bluetoothAdapter!!.isEnabled) {
            bluetoothAdapter!!.enable()
        }
        if (bluetoothAdapter!!.scanMode != BluetoothAdapter.SCAN_MODE_CONNECTABLE_DISCOVERABLE) {
            val discoveryIntent = Intent(BluetoothAdapter.ACTION_REQUEST_DISCOVERABLE)
            discoveryIntent.putExtra(BluetoothAdapter.EXTRA_DISCOVERABLE_DURATION, 300)
            startActivity(discoveryIntent)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (chatClass != null) {
            chatClass!!.stop()
        }
    }

    companion object {
        const val MESSAGE_STATE_CHANGED = 0
        const val MESSAGE_READ = 1
        const val MESSAGE_WRITE = 2
        const val MESSAGE_DEVICE_NAME = 3
        const val MESSAGE_TOAST = 4
        const val DEVICE_NAME = "device_name"
        const val TOAST = "toast"
    }
}